radixdlt/src/integration/java/com/radixdlt/consensus/simulation/invariants/bft/SafetyInvariant.java

```diff
@@ -91,48 +85,40 @@
-				final AtomicReference<View> highest = epochHighest.computeIfAbsent(epoch, e -> new AtomicReference<>());
-				final Map<BFTNode, View> lastCommittedByNode = epochLastCommittedByNode.computeIfAbsent(epoch, e -> new ConcurrentHashMap<>());
+				final EpochView epochView = EpochView.of(vertex.getEpoch(), vertex.getView());
 
-				final Vertex currentVertexAtView = committedVertices.get(vertex.getView());
+				final Vertex currentVertexAtView = committedVertices.get(epochView);
 				if (currentVertexAtView != null) {
 					if (!currentVertexAtView.getId().equals(vertex.getId())) {
 						return conflictingVerticesError(vertex, currentVertexAtView);
 					}
 				} else {
-					if (!vertex.getParentMetadata().getView().isGenesis()) {
-						View highestCommittedView = highest.get();
-						if (highestCommittedView == null) {
-							return badParentExpectedGenesisError(vertex);
-						}
-
-						final Vertex lastCommitted = committedVertices.get(highestCommittedView);
-						if (vertex.getParentId() == null) {
-							return noParentError(vertex);
-						}
-
-						if (!vertex.getParentId().equals(lastCommitted.getId())
-							|| !vertex.getParentMetadata().getView().equals(lastCommitted.getView())) {
-							return badParentError(vertex, lastCommitted);
+					EpochView parentEpochView = EpochView.of(vertex.getEpoch(), vertex.getParentMetadata().getView());
+					Vertex parent = committedVertices.get(parentEpochView);
+					if (parent == null) {
+						Entry<EpochView, Vertex> higherCommitted = committedVertices.higherEntry(parentEpochView);
+						if (higherCommitted != null) {
+							VertexMetadata higherParentMetadata = higherCommitted.getValue().getParentMetadata();
+							EpochView higherCommittedParentEpochView = EpochView.of(higherParentMetadata.getEpoch(), higherParentMetadata.getView());
+							if (epochView.compareTo(higherCommittedParentEpochView) > 0) {
+								return brokenChainError(vertex, higherCommitted.getValue());
+							}
 						}
 					}
 
-					committedVertices.put(vertex.getView(), vertex);
-					highest.set(vertex.getView());
+					committedVertices.put(epochView, vertex);
 				}
 
 				// Clean up old vertices so that we avoid consuming too much memory
-				lastCommittedByNode.put(node, vertex.getView());
-				final View lowest = network.getNodes().stream()
-					.map(n -> lastCommittedByNode.getOrDefault(n, View.genesis()))
+				lastCommittedByNode.put(node, epochView);
+				final EpochView lowest = network.getNodes().stream()
+					.map(n -> lastCommittedByNode.getOrDefault(n, EpochView.of(0, View.genesis())))
 					.reduce((v0, v1) -> v0.compareTo(v1) < 0 ? v0 : v1)
-					.orElse(View.genesis());
-				final Set<View> viewsToRemove = committedVertices.keySet().stream()
+					.orElse(EpochView.of(0, View.genesis()));
+				final Set<EpochView> viewsToRemove = committedVertices.keySet().stream()
```

Comment:
Could replace this with `committedVertices.headMap(lowest).clear()`.

Reply:
ooo nice